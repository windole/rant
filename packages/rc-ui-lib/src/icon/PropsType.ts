import React from 'react';
import { BadgeSettingProps, HtmlTagType } from '../badge/PropsType';
import { BaseTypeProps } from '../utils';

export interface IconProps extends BaseTypeProps {
  /** HTML 标签
   * @default 'i'
   */
  tag?: HtmlTagType;
  /** 图标名称或图片链接   */
  name: string;
  /** 图标颜色   */
  color?: string;
  /** 是否显示图标右上角小红点 */
  dot?: boolean;
  /**
   * 图标大小，如 20px 2em，默认单位为px
   * @default 'inherit''
   */
  size?: number | string;
  /** 徽标内容 */
  badge?: BadgeSettingProps;
  /**
   * 类名前缀，用于使用自定义图标
   * @default 'van-icon'
   */
  classPrefix?: string;
  /** 自元素 */
  children?: React.ReactNode;
  /** 点击图标时触发   */
  onClick?: (e: React.MouseEvent) => void;
  /** 触摸开始，多点触控 */
  onTouchStart?: (e: React.MouseEvent) => void;
}
