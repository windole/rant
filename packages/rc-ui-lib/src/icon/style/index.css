.van-icon__image {
  width: 1em;
  height: 1em;
  object-fit: contain;
}
