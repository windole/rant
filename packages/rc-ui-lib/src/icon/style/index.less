@import '@vant/icons/src/encode-woff2.less';

.rc-icon {
  &__image {
    width: 1em;
    height: 1em;
    object-fit: contain;
  }
}
