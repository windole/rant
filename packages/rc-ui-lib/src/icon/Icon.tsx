import React, { useContext, useMemo } from 'react';
import classnames from 'classnames';
import { addUnit } from '../utils';
import { IconProps } from './PropsType';
import Badge from '../badge';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';

function isImage(name?: string): boolean {
  return name ? name.indexOf('/') !== -1 : false;
}

const Icon: React.FC<IconProps> = (props) => {
  const { iconPrefix } = useContext(ConfigProviderContext);

  const { dot, tag, name, className, onClick, onTouchStart } = props;

  const imageIcon = isImage(name);

  const classPrefix = useMemo(
    () => props.classPrefix || iconPrefix,
    [props.classPrefix, iconPrefix],
  );

  return (
    <Badge
      dot={dot}
      tag={tag}
      className={classnames(className, classPrefix, imageIcon ? '' : `${classPrefix}-${name}`)}
      style={{
        color: props.color,
        fontSize: addUnit(props.size),
        ...props.style,
      }}
      onClick={onClick}
      onTouchStart={onTouchStart}
      {...props.badge}
    >
      {props?.children}
      {imageIcon && <img className={classnames('rc-icon__image')} src={name} alt={name} />}
    </Badge>
  );
};

Icon.defaultProps = {
  tag: 'i',
};

export default Icon;
