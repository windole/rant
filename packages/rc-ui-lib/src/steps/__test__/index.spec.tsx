import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Steps } from '..';
import { Icon } from '../../icon';

describe('Steps', () => {
  let wrapper;
  afterEach(() => {
    wrapper?.unmount();
    jest.restoreAllMocks();
  });

  test('should render correctly', () => {
    wrapper = mount(
      <Steps active={0}>
        <Steps.Item>B</Steps.Item>
        <Steps.Item>A</Steps.Item>
      </Steps>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should render custom icon correctly', () => {
    wrapper = mount(
      <Steps
        active={1}
        activeIcon={<Icon name="success" />}
        finishIcon={<Icon name="finish" />}
        inactiveIcon={<Icon name="fail" />}
      >
        <Steps.Item>B</Steps.Item>
        <Steps.Item>A</Steps.Item>
        <Steps.Item>C</Steps.Item>
      </Steps>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should emit click-step event when step is clicked', () => {
    const onClickStep = jest.fn();
    wrapper = mount(
      <Steps active={1} onClickStep={onClickStep}>
        <Steps.Item>A</Steps.Item>
        <Steps.Item>B</Steps.Item>
        <Steps.Item>C</Steps.Item>
      </Steps>,
    );

    wrapper.find('.rc-step').at(0).simulate('click');
    expect(onClickStep).toHaveBeenCalledTimes(0);

    wrapper.find('.rc-step__title').at(0).simulate('click');
    expect(onClickStep).toHaveBeenCalledWith(0);

    wrapper.find('.rc-step__circle-container').at(2).simulate('click');
    expect(onClickStep).toHaveBeenCalledTimes(2);
    expect(onClickStep).toHaveBeenLastCalledWith(2);
  });

  test('should change inactive color when using inactive-color prop', () => {
    wrapper = mount(
      <Steps active={0} inactiveColor="red">
        <Steps.Item>A</Steps.Item>
        <Steps.Item>B</Steps.Item>
      </Steps>,
    );
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should change inactive icon when using inactive-icon prop', () => {
    wrapper = mount(
      <Steps active={0} inactiveIcon="foo">
        <Steps.Item>A</Steps.Item>
        <Steps.Item>B</Steps.Item>
      </Steps>,
    );

    const steps = wrapper.find('.rc-step').at(1);
    expect(steps.find('.van-icon-foo').exists()).toBeTruthy();
    expect(steps.html()).toMatchSnapshot();
  });

  test('should change finish icon when using finish-icon prop', () => {
    wrapper = mount(
      <Steps active={1} finishIcon="foo">
        <Steps.Item>A</Steps.Item>
        <Steps.Item>B</Steps.Item>
      </Steps>,
    );

    const firstStep = wrapper.find('.rc-step');
    expect(firstStep.find('.van-icon-foo').exists()).toBeTruthy();
    expect(toJson(firstStep)).toMatchSnapshot();
  });

  test('should render icon-prefix correctly', () => {
    wrapper = mount(
      <Steps active={1} iconPrefix="custom-icon">
        <Steps.Item>A</Steps.Item>
        <Steps.Item>B</Steps.Item>
      </Steps>,
    );

    const steps = wrapper.find('.rc-step').at(1);
    expect(toJson(steps)).toMatchSnapshot();
  });
});
