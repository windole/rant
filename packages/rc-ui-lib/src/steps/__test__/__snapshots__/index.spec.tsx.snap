// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Steps should change finish icon when using finish-icon prop 1`] = `
Array [
  <div
    className="rc-hairline rc-step rc-step--horizontal rc-step--finish"
  >
    <div
      className="rc-step__title"
      onClick={[Function]}
      style={Object {}}
    >
      A
    </div>
    <div
      className="rc-step__circle-container"
      onClick={[Function]}
    >
      <Icon
        className="rc-step__icon rc-step__icon--finish"
        name="foo"
        tag="i"
      >
        <Badge
          className="rc-step__icon rc-step__icon--finish van-icon van-icon-foo"
          showZero={true}
          style={
            Object {
              "color": undefined,
              "fontSize": undefined,
            }
          }
          tag="i"
        >
          <i
            className="rc-badge__wrapper rc-step__icon rc-step__icon--finish van-icon van-icon-foo"
            style={
              Object {
                "color": undefined,
                "fontSize": undefined,
              }
            }
          />
        </Badge>
      </Icon>
    </div>
    <div
      className="rc-step__line"
      style={
        Object {
          "background": undefined,
        }
      }
    />
  </div>,
  <div
    className="rc-hairline rc-step rc-step--horizontal rc-step--process"
  >
    <div
      className="rc-step__title rc-step__title--active"
      onClick={[Function]}
      style={
        Object {
          "color": undefined,
        }
      }
    >
      B
    </div>
    <div
      className="rc-step__circle-container"
      onClick={[Function]}
    >
      <Icon
        className="rc-step__icon rc-step__icon--active"
        name="checked"
        tag="i"
      >
        <Badge
          className="rc-step__icon rc-step__icon--active van-icon van-icon-checked"
          showZero={true}
          style={
            Object {
              "color": undefined,
              "fontSize": undefined,
            }
          }
          tag="i"
        >
          <i
            className="rc-badge__wrapper rc-step__icon rc-step__icon--active van-icon van-icon-checked"
            style={
              Object {
                "color": undefined,
                "fontSize": undefined,
              }
            }
          />
        </Badge>
      </Icon>
    </div>
    <div
      className="rc-step__line"
      style={
        Object {
          "background": undefined,
        }
      }
    />
  </div>,
]
`;

exports[`Steps should change inactive color when using inactive-color prop 1`] = `"<div class=\\"rc-steps rc-steps--horizontal\\"><div class=\\"rc-steps__items\\"><div class=\\"rc-hairline rc-step rc-step--horizontal rc-step--process\\"><div class=\\"rc-step__title rc-step__title--active\\">A</div><div class=\\"rc-step__circle-container\\"><i class=\\"rc-badge__wrapper rc-step__icon rc-step__icon--active van-icon van-icon-checked\\"></i></div><div class=\\"rc-step__line\\" style=\\"background: red;\\"></div></div><div class=\\"rc-hairline rc-step rc-step--horizontal rc-step--waiting\\"><div class=\\"rc-step__title\\" style=\\"color: red;\\">B</div><div class=\\"rc-step__circle-container\\"><i class=\\"rc-step__circle\\" style=\\"background: red;\\"></i></div><div class=\\"rc-step__line\\" style=\\"background: red;\\"></div></div></div></div>"`;

exports[`Steps should change inactive icon when using inactive-icon prop 1`] = `"<div class=\\"rc-hairline rc-step rc-step--horizontal rc-step--waiting\\"><div class=\\"rc-step__title\\">B</div><div class=\\"rc-step__circle-container\\"><i class=\\"rc-badge__wrapper rc-step__icon van-icon van-icon-foo\\"></i></div><div class=\\"rc-step__line\\"></div></div>"`;

exports[`Steps should render correctly 1`] = `
<Steps
  active={0}
  activeIcon="checked"
  direction="horizontal"
>
  <div
    className="rc-steps rc-steps--horizontal"
  >
    <div
      className="rc-steps__items"
    >
      <StepsItem
        index={0}
        key=".0"
        parent={
          Object {
            "active": 0,
            "activeIcon": "checked",
            "direction": "horizontal",
          }
        }
      >
        <div
          className="rc-hairline rc-step rc-step--horizontal rc-step--process"
        >
          <div
            className="rc-step__title rc-step__title--active"
            onClick={[Function]}
            style={
              Object {
                "color": undefined,
              }
            }
          >
            B
          </div>
          <div
            className="rc-step__circle-container"
            onClick={[Function]}
          >
            <Icon
              className="rc-step__icon rc-step__icon--active"
              name="checked"
              tag="i"
            >
              <Badge
                className="rc-step__icon rc-step__icon--active van-icon van-icon-checked"
                showZero={true}
                style={
                  Object {
                    "color": undefined,
                    "fontSize": undefined,
                  }
                }
                tag="i"
              >
                <i
                  className="rc-badge__wrapper rc-step__icon rc-step__icon--active van-icon van-icon-checked"
                  style={
                    Object {
                      "color": undefined,
                      "fontSize": undefined,
                    }
                  }
                />
              </Badge>
            </Icon>
          </div>
          <div
            className="rc-step__line"
            style={
              Object {
                "background": undefined,
              }
            }
          />
        </div>
      </StepsItem>
      <StepsItem
        index={1}
        key=".1"
        parent={
          Object {
            "active": 0,
            "activeIcon": "checked",
            "direction": "horizontal",
          }
        }
      >
        <div
          className="rc-hairline rc-step rc-step--horizontal rc-step--waiting"
        >
          <div
            className="rc-step__title"
            onClick={[Function]}
            style={
              Object {
                "color": undefined,
              }
            }
          >
            A
          </div>
          <div
            className="rc-step__circle-container"
            onClick={[Function]}
          >
            <i
              className="rc-step__circle"
              style={
                Object {
                  "background": undefined,
                }
              }
            />
          </div>
          <div
            className="rc-step__line"
            style={
              Object {
                "background": undefined,
              }
            }
          />
        </div>
      </StepsItem>
    </div>
  </div>
</Steps>
`;

exports[`Steps should render custom icon correctly 1`] = `
<Steps
  active={1}
  activeIcon={
    <Icon
      name="success"
      tag="i"
    />
  }
  direction="horizontal"
  finishIcon={
    <Icon
      name="finish"
      tag="i"
    />
  }
  inactiveIcon={
    <Icon
      name="fail"
      tag="i"
    />
  }
>
  <div
    className="rc-steps rc-steps--horizontal"
  >
    <div
      className="rc-steps__items"
    >
      <StepsItem
        index={0}
        key=".0"
        parent={
          Object {
            "active": 1,
            "activeIcon": <Icon
              name="success"
              tag="i"
            />,
            "direction": "horizontal",
            "finishIcon": <Icon
              name="finish"
              tag="i"
            />,
            "inactiveIcon": <Icon
              name="fail"
              tag="i"
            />,
          }
        }
      >
        <div
          className="rc-hairline rc-step rc-step--horizontal rc-step--finish"
        >
          <div
            className="rc-step__title"
            onClick={[Function]}
            style={Object {}}
          >
            B
          </div>
          <div
            className="rc-step__circle-container"
            onClick={[Function]}
          >
            <Icon
              name="finish"
              tag="i"
            >
              <Badge
                className="van-icon van-icon-finish"
                showZero={true}
                style={
                  Object {
                    "color": undefined,
                    "fontSize": undefined,
                  }
                }
                tag="i"
              >
                <i
                  className="rc-badge__wrapper van-icon van-icon-finish"
                  style={
                    Object {
                      "color": undefined,
                      "fontSize": undefined,
                    }
                  }
                />
              </Badge>
            </Icon>
          </div>
          <div
            className="rc-step__line"
            style={
              Object {
                "background": undefined,
              }
            }
          />
        </div>
      </StepsItem>
      <StepsItem
        index={1}
        key=".1"
        parent={
          Object {
            "active": 1,
            "activeIcon": <Icon
              name="success"
              tag="i"
            />,
            "direction": "horizontal",
            "finishIcon": <Icon
              name="finish"
              tag="i"
            />,
            "inactiveIcon": <Icon
              name="fail"
              tag="i"
            />,
          }
        }
      >
        <div
          className="rc-hairline rc-step rc-step--horizontal rc-step--process"
        >
          <div
            className="rc-step__title rc-step__title--active"
            onClick={[Function]}
            style={
              Object {
                "color": undefined,
              }
            }
          >
            A
          </div>
          <div
            className="rc-step__circle-container"
            onClick={[Function]}
          >
            <Icon
              name="success"
              tag="i"
            >
              <Badge
                className="van-icon van-icon-success"
                showZero={true}
                style={
                  Object {
                    "color": undefined,
                    "fontSize": undefined,
                  }
                }
                tag="i"
              >
                <i
                  className="rc-badge__wrapper van-icon van-icon-success"
                  style={
                    Object {
                      "color": undefined,
                      "fontSize": undefined,
                    }
                  }
                />
              </Badge>
            </Icon>
          </div>
          <div
            className="rc-step__line"
            style={
              Object {
                "background": undefined,
              }
            }
          />
        </div>
      </StepsItem>
      <StepsItem
        index={2}
        key=".2"
        parent={
          Object {
            "active": 1,
            "activeIcon": <Icon
              name="success"
              tag="i"
            />,
            "direction": "horizontal",
            "finishIcon": <Icon
              name="finish"
              tag="i"
            />,
            "inactiveIcon": <Icon
              name="fail"
              tag="i"
            />,
          }
        }
      >
        <div
          className="rc-hairline rc-step rc-step--horizontal rc-step--waiting"
        >
          <div
            className="rc-step__title"
            onClick={[Function]}
            style={
              Object {
                "color": undefined,
              }
            }
          >
            C
          </div>
          <div
            className="rc-step__circle-container"
            onClick={[Function]}
          >
            <Icon
              name="fail"
              tag="i"
            >
              <Badge
                className="van-icon van-icon-fail"
                showZero={true}
                style={
                  Object {
                    "color": undefined,
                    "fontSize": undefined,
                  }
                }
                tag="i"
              >
                <i
                  className="rc-badge__wrapper van-icon van-icon-fail"
                  style={
                    Object {
                      "color": undefined,
                      "fontSize": undefined,
                    }
                  }
                />
              </Badge>
            </Icon>
          </div>
          <div
            className="rc-step__line"
            style={
              Object {
                "background": undefined,
              }
            }
          />
        </div>
      </StepsItem>
    </div>
  </div>
</Steps>
`;

exports[`Steps should render icon-prefix correctly 1`] = `
<div
  className="rc-hairline rc-step rc-step--horizontal rc-step--process"
>
  <div
    className="rc-step__title rc-step__title--active"
    onClick={[Function]}
    style={
      Object {
        "color": undefined,
      }
    }
  >
    B
  </div>
  <div
    className="rc-step__circle-container"
    onClick={[Function]}
  >
    <Icon
      className="rc-step__icon rc-step__icon--active"
      classPrefix="custom-icon"
      name="checked"
      tag="i"
    >
      <Badge
        className="rc-step__icon rc-step__icon--active custom-icon custom-icon-checked"
        showZero={true}
        style={
          Object {
            "color": undefined,
            "fontSize": undefined,
          }
        }
        tag="i"
      >
        <i
          className="rc-badge__wrapper rc-step__icon rc-step__icon--active custom-icon custom-icon-checked"
          style={
            Object {
              "color": undefined,
              "fontSize": undefined,
            }
          }
        />
      </Badge>
    </Icon>
  </div>
  <div
    className="rc-step__line"
    style={
      Object {
        "background": undefined,
      }
    }
  />
</div>
`;
